### ISSUE TITLE
[Bug]: typecheck bug on the DataTableFacetedFilter example

### ISSUE THREAD

**Issue by maelp** [radix-vue/shadcn-vue#327]
### Reproduction

None

### Describe the bug

When using the code from the Task table example DataTableFacetedFilter I have the following issue

<img width="628" alt="image" src="https://github.com/radix-vue/shadcn-vue/assets/15458/83b2a150-db12-40f7-9360-608a120d7c97">

```
<Command
        :filter-function="
          (list: DataTableFacetedFilter['options'], term) =>
            list.filter((i) => i.title.toLowerCase()?.includes(term))
        "
      >
```

I'm using the latest radix-vue, but it seems that filter-function first parameter should use "AcceptableValue[]" list, which contains string, number, object etc, but it doesn't typecheck for an arbitrary object for some reason(?)

### System Info

```bash
System:
    OS: macOS 14.1.2
    CPU: (8) arm64 Apple M1
    Memory: 95.53 MB / 8.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.10.0 - ~/.nvm/versions/node/v20.10.0/bin/node
    Yarn: 1.16.0 - ~/.node/bin/yarn
    npm: 10.2.3 - ~/.nvm/versions/node/v20.10.0/bin/npm
    pnpm: 8.15.0 - ~/Library/pnpm/pnpm
    Watchman: 2024.01.22.00 - /opt/homebrew/bin/watchman
  Browsers:
    Chrome: 121.0.6167.139
    Safari: 17.1.2
  npmPackages:
    @vueuse/core: 10.7.1 => 10.7.1
    radix-vue: 1.4.0 => 1.4.0
    vue: 3.4.7 => 3.4.7
```

### Contributes

- [ ] I am willing to submit a PR to fix this issue
- [ ] I am willing to submit a PR with failing tests

**Comment by hrynevychroman** [comment]
@maelp can you hover over underlined red text and show error too, please)

**Comment by maelp** [comment]
<img width="662" alt="image" src="https://github.com/radix-vue/shadcn-vue/assets/15458/eff28283-f3c7-41d3-ad4b-bcff0cf5d8a9">

**Comment by maelp** [comment]
<img width="656" alt="image" src="https://github.com/radix-vue/shadcn-vue/assets/15458/01f66f3a-6265-4ab5-9a46-5ea93786c80e">

**Comment by hrynevychroman** [comment]
thanks, will take a look on it)

**Comment by maelp** [comment]
BTW adding ":string" to term doesn't fix the bug for me (taking a look at your PR)

**Comment by maelp** [comment]
<img width="627" alt="image" src="https://github.com/radix-vue/shadcn-vue/assets/15458/212e61c0-dc43-4ed8-a99b-1f1aff63adda">

**Comment by hrynevychroman** [comment]
Yes, it just little issue so never mind, I will remove `draft` status when resolve it)
https://github.com/radix-vue/shadcn-vue/issues/327#issuecomment-1935001998

**Comment by hrynevychroman** [comment]
@maelp To fix this issue you need to define `Option` interface globally and pass it inside `ComboboxRootProps` in `Command.vue`, this will remove error inside your editor

<img width="1016" alt="image" src="https://github.com/radix-vue/shadcn-vue/assets/82209198/69ac17d3-8679-4606-9882-ee03fa4794af">

<img width="710" alt="image" src="https://github.com/radix-vue/shadcn-vue/assets/82209198/a0c18093-930c-4c49-9137-0d1e7e84257a">

<img width="1033" alt="image" src="https://github.com/radix-vue/shadcn-vue/assets/82209198/2e084885-58cc-429c-9af1-cf54eb9107bc">


Need to think about how to fix it globally, because issue is inside radix-vue components, when you don't define `T` it takes default value of `AcceptableValue` which is 

```typescript
export type AcceptableValue = string | number | boolean | object;
```

It's logical that `Option` extends `object`, but VS Code internal TS Server don't think so 😞 

Will try to think with it, maybe need to fix radix-vue TS types 🤔 

@sadeghbarati maybe you will look and have some ideas, I will appreciate it ❤️

**Comment by maelp** [comment]
Interesting... indeed I didn't know whether the issue was that the method is expecting any AcceptableValue and typescript is complaining that we are only handling the case where the filterFunction parameter is of type `DataTableFacetedFilter['option']` and not the other cases that it thinks could happen (eg string, number, etc)... wondering if that's the issue? in that case it should be generic?

**Comment by hrynevychroman** [comment]
No, issue is that typescript don't think that our custom object for option is `object` 😁 

In my codebase I never use `object` definition, because it not just Objects. Functions, Arrays is also type of `object` is JS, think that maybe replace `object` to type `Record<string, any>` will help with this 🤔

**Comment by tuanalumi** [comment]
Is this fixed? I am still getting the type error on `filterFunction`. Not sure how to get it right..

<img width="949" alt="image" src="https://github.com/radix-vue/shadcn-vue/assets/5226413/91a80333-188e-4e03-9a7f-79017bdbcb1a">

**Comment by 9mm** [comment]
I'm also getting the same issue...

```typescript
export type FilterItem = {
  value: string;
  label: string;
};

const onSift = (list: FilterItem[], term: string) => {
  return list.filter(item => item.label.toLowerCase().includes(term.toLowerCase()));
};

Diagnostics:
1. Type '(list: FilterItem[], term: string) => globalThis.FilterItem[]' is not assignable to type '(val: string[] | number[] | false[] | true[] | Record<string, any>[], term: string) => string[] | number[] | false[] | true[] | Record<string, any>[]'. [2322]
```

**Comment by 9mm** [comment]
> @maelp To fix this issue you need to define `Option` interface globally and pass it inside `ComboboxRootProps` in `Command.vue`, this will remove error inside your editor
> 
> <img alt="image" width="1016" src="https://private-user-images.githubusercontent.com/82209198/303501376-69ac17d3-8679-4606-9882-ee03fa4794af.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzNjU2NzYsIm5iZiI6MTcxOTM2NTM3NiwicGF0aCI6Ii84MjIwOTE5OC8zMDM1MDEzNzYtNjlhYzE3ZDMtODY3OS00NjA2LTk4ODItZWUwM2ZhNDc5NGFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI2VDAxMjkzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJmNzUwNTg0ZWI4ZWJhOTI1MzQ1MWJlNWY1ZTgwNzZkYmI2ZmRmYmY1NjhkNWY0MzFhZDEzYzVkMmY4ZTNhOTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OHTKCCvLhxMvcOncv4JpzVYxwZ1hzUtqqI8zntLEazQ"> <img alt="image" width="710" src="https://private-user-images.githubusercontent.com/82209198/303501388-a0c18093-930c-4c49-9137-0d1e7e84257a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzNjU2NzYsIm5iZiI6MTcxOTM2NTM3NiwicGF0aCI6Ii84MjIwOTE5OC8zMDM1MDEzODgtYTBjMTgwOTMtOTMwYy00YzQ5LTkxMzctMGQxZTdlODQyNTdhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI2VDAxMjkzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU4YWQ5YjczYWFlYmFjNmQ0YWNhOWFkN2ZiNzBmMDVjZWIyNTExZDg2ZWQzODg5MTJhNzUxZTg2NThkYTNkMTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3KM6-GWB5GT5oWTBDqdn6fdiUarX5I-sZOlh-evqo3Q"> <img alt="image" width="1033" src="https://private-user-images.githubusercontent.com/82209198/303501405-2e084885-58cc-429c-9af1-cf54eb9107bc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzNjU2NzYsIm5iZiI6MTcxOTM2NTM3NiwicGF0aCI6Ii84MjIwOTE5OC8zMDM1MDE0MDUtMmUwODQ4ODUtNThjYy00MjljLTlhZjEtY2Y1NGViOTEwN2JjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI2VDAxMjkzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU5MDUxOWNhZDFlYzk1MTIwNzk2OTc5YjZmYmZlZmQxNDhiMDRhZWUyZGRhMWE4YWY3YWMzNGMxYTYxOTY4N2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V6Xp1gjTwCjjG2tJMfR0zy6iJWppNjrBbzuFBjiNyrM">
> Need to think about how to fix it globally, because issue is inside radix-vue components, when you don't define `T` it takes default value of `AcceptableValue` which is
> 
> ```ts
> export type AcceptableValue = string | number | boolean | object;
> ```
> 
> It's logical that `Option` extends `object`, but VS Code internal TS Server don't think so 😞
> 
> Will try to think with it, maybe need to fix radix-vue TS types 🤔
> 
> @sadeghbarati maybe you will look and have some ideas, I will appreciate it ❤️

@romanhrynevych so is the other part of this solution commenting out `modelValue` still correct, or was that for debugging? I dont want to break the component with unexpected side effects, but otherwise I get:

> Diagnostics:
1. Type 'string' is not assignable to type '(props: LooseRequired<ComboboxRootProps<FilterItem> & { class?: any; }>) => FilterItem | FilterItem[]'. [2322]

**Comment by kratos-digital** [comment]
> @maelp To fix this issue you need to define `Option` interface globally and pass it inside `ComboboxRootProps` in `Command.vue`, this will remove error inside your editor
> 
> <img alt="image" width="1016" src="https://private-user-images.githubusercontent.com/82209198/303501376-69ac17d3-8679-4606-9882-ee03fa4794af.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY0OTc4NDEsIm5iZiI6MTcyNjQ5NzU0MSwicGF0aCI6Ii84MjIwOTE5OC8zMDM1MDEzNzYtNjlhYzE3ZDMtODY3OS00NjA2LTk4ODItZWUwM2ZhNDc5NGFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE2VDE0MzkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFlYjc1MjliZGVkNTVmYmQ2ZmY0ZGU3ZTc5MWE4YTUxNTkxODQxYzVhNTljYjI4OTA5MThmOTlkMzc4OTgwOTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.clvihdNhC5U3fhCMJLLLlWRNFtyAv9m-u624q6fia7s"> <img alt="image" width="710" src="https://private-user-images.githubusercontent.com/82209198/303501388-a0c18093-930c-4c49-9137-0d1e7e84257a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY0OTc4NDEsIm5iZiI6MTcyNjQ5NzU0MSwicGF0aCI6Ii84MjIwOTE5OC8zMDM1MDEzODgtYTBjMTgwOTMtOTMwYy00YzQ5LTkxMzctMGQxZTdlODQyNTdhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE2VDE0MzkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ4MzAyZWI2ODM1NjYxNjZmZGNhMjZiZTNhYzVkYTU0ZDUwMzBhMmYxYWEzOGNiNTNhMzRlMjMzYmFkMTAzMzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cL8RC6zy4hZbjKeH1LPDg_vDrqNst06VTd6do6mu2NU"> <img alt="image" width="1033" src="https://private-user-images.githubusercontent.com/82209198/303501405-2e084885-58cc-429c-9af1-cf54eb9107bc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY0OTc4NDEsIm5iZiI6MTcyNjQ5NzU0MSwicGF0aCI6Ii84MjIwOTE5OC8zMDM1MDE0MDUtMmUwODQ4ODUtNThjYy00MjljLTlhZjEtY2Y1NGViOTEwN2JjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE2VDE0MzkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1MmI5OWYwYzJmYTlhYzliMDBiMDkxNmMwODNjZDQyZTAzYzU1NTZlZWE4ZTY3MzZlNTFjZjgyOTM0NTc5NzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BRry09aTxMsFBAko_FPXUDUJDeNYL3cy0BsvnRAxSvU">
> Need to think about how to fix it globally, because issue is inside radix-vue components, when you don't define `T` it takes default value of `AcceptableValue` which is
> 
> ```ts
> export type AcceptableValue = string | number | boolean | object;
> ```
> 
> It's logical that `Option` extends `object`, but VS Code internal TS Server don't think so 😞
> 
> Will try to think with it, maybe need to fix radix-vue TS types 🤔
> 
> @sadeghbarati maybe you will look and have some ideas, I will appreciate it ❤️

This fixed the error for me.

In types/index.d.ts
```
export interface Option {
  label: string;
  value: string;
  icon?: Component;
}
```

In Command.vue
```
import { Option } from "@/types";

const props = withDefaults(
  defineProps<
    ComboboxRootProps<Option> & { class?: HTMLAttributes["class"] }
  >(),
  {
    open: true,
  },
);
```

In FacetedFilter.vue (or wherever you define your filters)
```
interface DataTableFacetedFilter {
  column: Column<any, any>;
  title: string;
  options: Option[];
  default?: string | boolean;
  search?: boolean;
}
```

```
<Command
        :filter-function="
          (list: DataTableFacetedFilter['options'], term: string) =>
            list.filter((row) => row.label.toLowerCase()?.includes(term))
        "
      >
```

**Comment by sadeghbarati** [comment]
Error seems fixed with new `reka-ui` components
